Modules/_abc.c

```
@@ -814,23 +839,49 @@ _abc__abc_subclasscheck_impl(PyObject *module, PyObject *self,
         if (scls == NULL) {
             goto end;
         }
+
+        _abc_data *scls_impl;
+        int scls_is_abc = _get_impl_optional(module, scls, &scls_impl);
+        if (scls_is_abc < 0) {
+            goto end;
+        }
+        if (scls_is_abc > 0) {
+            /*
+              If inside recursive issubclass check, avoid adding classes
+              to any cache because this may drastically increase memory usage.
+              Unfortunately, issubclass/__subclasscheck__ don't accept third
+              argument with context, so using global context within ABCMeta.
+              This is done only on first method call, next will use cache.
+            */
+            set_issubclasscheck_recursive(scls_impl);
+        }
+
         int r = PyObject_IsSubclass(subclass, scls);
         Py_DECREF(scls);
+        if (scls_is_abc > 0) {
+            unset_issubclasscheck_recursive(scls_impl);
+            Py_DECREF(scls_impl);
+        }
+
+        if (r < 0) {
```

`r < 0` must be handled before `scls_is_abc > 0` as we could have a failure.

---

The whole point of having `scls_is_abc` and `scls_impl` is to enable recursion check before we enter `PyObject_IsSubclass` which can perform recursive `__subclasscheck__` call.

---

It doesn't matter. Errors should be handled before successful paths.

---

Please read my comment again, and the comment in the source code as well

---

No I mean, errors from `PyObject_IsSubclass` shouldn't be silently skipped. Usually, we can't call the C API if there's an exception set. If `PyObject_IsSubclass` fails, I don't think we should continue.

---

Fixed

---

> errors from PyObject_IsSubclass shouldn't be silently skipped (they could be fatal errors). > Usually, we can't call the C API if there's an exception set

Custom `__subclasscheck__` hook could raise an exception. In this case recursion check will not be reset.